File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/composite/CompositePattern.java
package jv.com.tencent.cain.design_pattern.composite;

/**
 * @author cainjiang
 * @date 2020/5/7
 */
public class CompositePattern {
    public static void main(String[] args) {
        Composite root = new Composite("root");
        root.add(new Leaf("Leaf A"));
        root.add(new Leaf("Leaf B"));

        Composite composite = new Composite("Composite X");
        composite.add(new Leaf("Leaf XA"));
        composite.add(new Leaf("Leaf XB"));
        root.add(composite);

        Composite composite1 = new Composite("Composite XY");
        composite1.add(new Leaf("Leaf XYA"));
        composite1.add(new Leaf("Leaf XYB"));
        root.add(composite1);

        root.add(new Leaf("Leaf C"));
        Leaf leaf = new Leaf("D");
        root.add(leaf);
        root.remove(leaf);

        root.display(1);
    }
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/composite/CompositePattern.kt
package kt.com.tencent.cain.desigin_pattern.composite


/**
 * @author cainjiang
 * @date 2020/5/7
 */
fun main() {
    val root = Composite("root")
    root.add(Leaf("Leaf A"))
    root.add(Leaf("Leaf B"))

    val composite = Composite("Composite X")
    composite.add(Leaf("Leaf XA"))
    composite.add(Leaf("Leaf XB"))
    root.add(composite)

    val composite1 = Composite("Composite XY")
    composite1.add(Leaf("Leaf XYA"))
    composite1.add(Leaf("Leaf XYB"))
    root.add(composite1)

    root.add(Leaf("Leaf C"))

    val leaf = Leaf("D")
    root.add(leaf)
    root.remove(leaf)

    root.display(1)
}